zerver/lib/users.py

```diff
@@ -511,6 +515,45 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
     return False
 
 
+def check_can_access_user(
+    target_user: UserProfile, user_profile: Optional[UserProfile] = None
+) -> bool:
+    if check_user_can_access_all_users(user_profile) or not user_access_restricted_in_realm(
+        target_user
+    ):
+        return True
+
+    assert user_profile is not None
+
+    if target_user.id == user_profile.id:
+        return True
+
+    # These include Subscription objects for streams as well as group DMs.
+    subscribed_recipient_ids = Subscription.objects.filter(
+        user_profile=user_profile,
+        active=True,
+        recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
+    ).values_list("recipient_id", flat=True)
+
+    if Subscription.objects.filter(
+        recipient_id__in=subscribed_recipient_ids, user_profile=target_user, active=True
+    ).exists():
+        return True
+
+    assert user_profile.recipient_id is not None
+    assert target_user.recipient_id is not None
+    direct_message_query = Message.objects.filter(
+        recipient__type=Recipient.PERSONAL, realm=target_user.realm
+    )
+    if direct_message_query.filter(
+        Q(sender_id=target_user.id, recipient_id=user_profile.recipient_id)
+        | Q(recipient_id=target_user.recipient_id, sender_id=user_profile.id)
+    ).exists():
+        return True
```

Comment:
Let's add a TODO comment somewhere around here that going to the `Message` table is expensive, so we do it last.

(Long-term, I think we will probably move 1:1 DMs to use the `Huddle` table, and thus be able to do this whole process using only the `Subscription`/`Recipient` queries above).